lighthouse-core/gather/driver.js

```diff
@@ -275,6 +276,98 @@ class Driver {
     this._nextProtocolTimeout = timeout;
   }
 
+  /**
+   * @param {LH.CrdpEvents['Target.receivedMessageFromTarget'][0]} event
+   * @param {string[]} parentSessionIds The list of session ids of the parents from oldest to youngest.
+   */
+  _handleReceivedMessageFromTarget(event, parentSessionIds) {
+    const {sessionId, message} = event;
+    /** @type {LH.Protocol.RawMessage} */
+    const protocolMessage = JSON.parse(message);
+
+    // Message was a response to some command, not an event, so we'll ignore it.
+    if ('id' in protocolMessage) return;
+
+    // We receive messages from the outermost subtarget which wraps the messages from the inner subtargets.
+    // We are recursively processing them from outside in, so build the list of parentSessionIds accordingly.
+    const sessionIds = parentSessionIds.concat([sessionId]);
+
+    if (protocolMessage.method === 'Target.receivedMessageFromTarget') {
+      // Unravel any messages from subtargets by recursively processing
+      this._handleReceivedMessageFromTarget(protocolMessage.params, sessionIds);
+    }
+
+    if (protocolMessage.method === 'Target.attachedToTarget') {
+      // Process any attachedToTarget messages from subtargets
+      this._handleTargetAttached(protocolMessage.params, sessionIds);
+    }
+
+    if (protocolMessage.method.startsWith('Network')) {
+      // Make sure the message gets added to our devtools log for replay later.
+      this._devtoolsLog.record(protocolMessage);
+      // Also dispatch the event to our status monitor for in-the-moment network quiet detection.
+      if (this._networkStatusMonitor) this._networkStatusMonitor.dispatch(protocolMessage);
+    }
+  }
+
+  /**
+   * @param {LH.CrdpEvents['Target.attachedToTarget'][0]} event
+   * @param {string[]} parentSessionIds The list of session ids of the parents from oldest to youngest.
+   */
+  _handleTargetAttached(event, parentSessionIds) {
+    const sessionIds = parentSessionIds.concat([event.sessionId]);
```

Comment:
`sessionIdPath` ? that'd be a bit more clear for me.

Reply:
yeah good call 👍